### PR TITLE
openssl: return to keg_only

### LINKED ISSUE #41613: OpenSSL linking issues due to removal of keg_only status
The homebrew OpenSSL recipe is no longer keg only (as of https://github.com/Homebrew/homebrew/commit/9ca3c054c070cc825870a6fc0fc5c5a22f7d8559#diff-5f8300a99f5c14b861dfa8eff2905a16) due to El Capitan no longer shipping headers.

Unfortunately, this leads to some major problems with projects (outside homebrew) that attempt to compile against OpenSSL without having the homebrew prefix added. Specifically, a project that links against OpenSSL and uses clang will see the homebrew OpenSSL headers due to the clang search path order:
```
#include <...> search starts here:
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.1.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
```

However, when it links the clang library search path order is:
```
Library search paths:
	/usr/lib
	/usr/local/lib
Framework search paths:
	/Library/Frameworks/
	/System/Library/Frameworks/
```

So the binary compiles using the headers for 1.0.2d but links against 0.9.8. This problem typically manifests as a Symbol Not Found error at runtime.

This problem can be worked around by passing the appropriate LDFLAGS or CFLAGS but appears to represent a significant regression from the previous behavior. Is there any better workaround or is this expected?

### PR DESCRIPTION
**Pull request by DomT4**
I'd like us to further investigate if this is necessary, but it looks like it might be, and since the bot is nice and quiet at the moment I've booted the timeout to giggle levels and will let this run through.

Closes #41613, if we merge it.

### CODE REVIEW

**Comment by DomT4** [comment]
@sjackman If we merge this, you may wish to merge an up-to-date Homebrew into Linuxbrew fairly quickly. See the linked issue for further detail; It'd be interesting to know if the problem impacts Linux as well as OS X.

**Comment by tdsmith** [comment]
openssl has never been keg-only in Linuxbrew since Linuxbrew is never installed to /usr/local.

**Comment by DomT4** [comment]
Oh, yup. Forgot Linuxbrew operates out of a different default prefix. Ignore the tired Dom.

**Comment by tdsmith** [comment]
The consequence of pulling this is breaking anything (other than these packages) which depends_on openssl and has been built from source since 9ca3c054c070cc was committed at 5pm Thursday July 9, UK time. I don't know how big that exposure is.

**Comment by DomT4** [comment]
Yeah. I double checked that earlier and there's breakage there for sure. I'm hoping with it being the weekend exposure will be minimal.

Worst-case scenario and we get a lot of failure reports: I was planning to deal with SSLv3 at some point in the not-completely-distant-future which would enforce a round of revisions, I could always bring that hastily forwards to deal with two breakages, one stone, etc.

**Comment by DomT4** [comment]
Sigh. This old problem has returned:

```
==> Installing wsmancli dependency: openwsman
==> Downloading https://homebrew.bintray.com/bottles/openwsman-2.6.0_1.mountain_lion.bottle.tar.gz
Error: Failed to download resource "openwsman"
Download failed: https://homebrew.bintray.com/bottles/openwsman-2.6.0_1.mountain_lion.bottle.tar.gz
Warning: Bottle installation failed: building from source.
Error: /usr/local/opt/automake not present or broken
Please reinstall automake. Sorry :(
```

Otherwise seems fine.

File: Library/Formula/openssl.rb [hunk 1]
```diff
@@ -17,6 +18,9 @@ class Openssl < Formula
 
   depends_on "makedepend" => :build
 
+  keg_only :provided_by_osx,
```
Comment (mikemcquaid) [inline Library/Formula/openssl.rb]: At the least this should be `keg_only :provided_pre_el_capitan` (the definition of which may need to be added) so this isn't keg-only there.

Comment (DomT4) [inline Library/Formula/openssl.rb]: @mistydemeo & @tdsmith seemed to raise objection to that in the chat - Will let them comment though.

**Comment by mikemcquaid** [comment]
@Homebrew/owners I'd like to hold off merging this for a couple of days (even if this means blocking updates to `openssl` dependents in that time). I don't think a single user report warrants changing this back (yet). I also think the linker/include issue is perhaps a pointer to another issue. The reason why it's picking up the wrong libs in this case is because of `libssl.dylib` rather than `libssl.1.0.0.dylib` linkage. Perhaps it would be enough/sensible to kill the `libssl.dylib` symlink?

Another related piece of data is that thoughtbot/laptop (which is popular enough and intersects with our user-base enough for us to see issues related to it) has been force-linking `openssl` for a while: https://github.com/thoughtbot/laptop/blob/61b093344a6c6c1051783aedfe9fab58145d9224/mac#L157

**Comment by DomT4** [comment]
If there's a way to not have to stuff OpenSSL back into its keg and at the same time ensure Clang doesn't break everything for giggles I'm all for it. I'd like to run the commands from the other thread in the 10.11 VM later today and see what the situation is there, It's probably the same but it'll help us decide on the 10.11 situation.

**Comment by mikemcquaid** [comment]
Another thought (before I jump on a long flight): perhaps the weird linking situation (and adding it to the caveats) is worth the security benefits for the OS X ecosystem. Regardless, we should definitely make this not keg-only on El Capitan.

**Comment by tdsmith** [comment]
I think it should remain keg-only on El Capitan. I think Misty was commenting on a different question.

Consider the case on El Capitan where a user builds something that `#include`s openssl/ssl.h and links with -lssl without further modifying the compiler search path. This cannot work, but the failure modes are different depending on whether openssl is keg-only.

If openssl is keg-only, the build fails because openssl/ssl.h can't be found. I think this is easy to understand and troubleshoot.

If openssl is linked, either: a) the build succeeds (with Homebrew's ssl configuration and a link to the system library, because /usr/lib is earlier in the library search path), which probably fails at runtime — this is particularly likely to be the case for extension modules or b) the build fails with an error about missing symbols.

I think the openssl-linked failure modes are harder to reason about if you don't know that OS X ships the openssl libraries but not the includes. The only benefit I see of linking openssl is making the `openssl` command more easily accessible, and maybe we can just link that in postinstall.

**Comment by mikemcquaid** [comment]
@tdsmith Yeh, fair enough. I still wonder why we've not been seeing issues with this when OpenSSL has been linked? Are people just linking against the old version and not noticing/caring? I'm coming around to the idea that we just keep this keg-only.

**Comment by DomT4** [comment]
Self-reminder: Add `bind` to the revision list here if/when we decide to go ahead with this.

**Comment by mikemcquaid** [comment]
@DomT4 Let's do it (whenever you get a chance).

**Comment by DomT4** [comment]
Cool. Will set a new build with bind going shortly, and merge it ASAP after that.

**Comment by sjackman** [comment]
@domt4 `keg_only :provided_by_osx` is ignored by Linuxbrew, but thanks for the heads up.

**Comment by DomT4** [comment]
@sjackman Yeah, Tim reminded me. Apologies for the false warning :)

**Comment by sjackman** [comment]
Linuxbrew definitely has issues with the include search path and the linker search path getting out of sync, particularly when packages in their build scripts do dumb things like e.g. add `/usr/lib` to the search path but don't add `/usr/include` to the search path in the same position. I never fully realized just how crazy this situation is of the header and library search paths being independently configurable. It creates all sorts of havoc.

**Comment by DomT4** [comment]
The only thing failing here is the unrelated bottle fetch for testing failure in https://github.com/Homebrew/homebrew/pull/41615#issuecomment-120687451. Pulling shortly.

**Pull request closed without merging**